autodiffcc/root.py

```diff
@@ -1,6 +1,110 @@
+    """
+    if start_values is None:
+        raise ValueError("Must include start_values as dict or list/array for this method.")
+
+    if isinstance(start_values, dict):
+        # turn keyword values into positional values matching function signature
+        values = []
+        for key in signature:
+            try:
+                values.append(start_values[key])
+            except KeyError:
+                raise KeyError(f"key {key} in function signature missing from start_values.")
+        if len(start_values.keys()) != len(signature):
+            raise KeyError("Too many keys passed in start_values dictionary.")
+    elif isinstance(start_values, (list, np.ndarray)):
+        # if list-like
+        values = start_values
+    elif np.isscalar(start_values):
+        values = [start_values]
+    else:
+        raise TypeError("Must include start_values as dict or list/array.")
+
+    values = np.array(values)
+
+    # check to make sure have correct number of variables
+    if len(values) != len(signature):
+        raise KeyError("Incorrect number of variables passed in start_values.")
+
+    return values
+
+
+def _check_interval(interval, signature):
+    """ A service that checks the shape of interval and reformats it for the specific method
+
+    INPUTS
+    =======
+    interval: List/array or a list of two dictionaries [interval_start, interval_end] provided to the root finder
+    signature: Signature of the function
+
+    RETURNS
+    ========
+    Returns two arrays interval_start, interval_end to pass to the specific root finding method
+    """
+
+    if interval is None:
+        raise ValueError("Must provide interval for this method.")
+
+    elif not isinstance(interval[0], type(interval[1])):
+        raise TypeError("Must include interval as list of two dicts or numeric list/array.")
+
+    elif isinstance(interval[0], dict) and isinstance(interval[1], dict):
+        # turn keyword values into positional values matching function signature
+        interval_start_values = []
+        start_dict = interval[0]
+
+        interval_end_values = []
+        end_dict = interval[1]
+
+        if len(start_dict.keys()) != len(end_dict.keys()):
+            raise KeyError("The interval_start and interval_end dictionaries must have the same number of keys.")
+
+        for key in signature:
+            try:
+                interval_start_values.append(start_dict[key])
+                interval_end_values.append(end_dict[key])
+            except KeyError:
+                raise KeyError(f"key {key} in function signature missing from interval dictionary.")
+        if len(start_dict.keys()) != len(signature):
+            raise KeyError("Too many keys passed in interval dictionary.")
+
+        interval_start = np.asarray(interval_start_values)
+        interval_end = np.asarray(interval_end_values)
+
+    elif isinstance(interval, (list, np.ndarray)):
+        # if list-like
+        values = np.asarray(interval)
```

Comment:
```
if len(values.shape) != 2:
     raise Error
```

Reply:
And then add a test for this

Reply:
Cool. Yeah that was the first approach I took, but didn't always work with other broken inputs, so I did two tests

```
    elif isinstance(interval, (list, np.ndarray)):
        # if list-like, check shape
        values = np.asarray(interval) # 2-D array, multivar
        if len(values.shape) == 2:
            interval_start = values[:, 0]
            interval_end = values[:, 1]
        elif len(interval) == 2: # 1-D Array, 1 var
            interval_start = np.asarray([values[0]])
            interval_end = np.asarray([values[1]])
        else:
            raise ValueError("Incorrect number of elements passed in interval.")
```